DataDog/dd-trace-rb
# [PROF-10679] Add preview support for correlating profiling with otel ruby gem

**What does this PR do?**

This PR picks up from where #3510 left off and adds preview support for correlating profiling data with traces/spans being tracked by the otel ruby gem (including endpoint profiling). (And specifically this PR just adds a few things on top of #3510, but does not otherwise change it significantly).

Note that this differs from the recommended way of using opentelemetry/otel with dd-trace-rb, which is to follow the instructions from https://docs.datadoghq.com/tracing/trace_collection/custom_instrumentation/otel_instrumentation/ruby/ .

The key difference is -- this PR makes code hotspots work even for setups that opt to not use `require 'datadog/opentelemetry'` (which we still recommend as the easier path).

The approach taken here is similar to #2342 and #3466: we peek inside the implementation of the otel gem to extract the information we need (namely the span id, local root span id, trace type, and trace endpoint). This approach is potentially brittle, which is why the code is written very defensively, with the aim of never breaking the application (or profiling) if something is off -- it just won't collect code hotspots.

This feature is off by default. This PR adds a new setting to control this feature:
* Via an environment variable (`DD_PROFILING_PREVIEW_OTEL_CONTEXT_ENABLED` -> set to `only`/`both`/`false`/`true`)
* Via code `c.profiling.advanced.preview_otel_context_enabled =` -> set to `"only"`/`"both"`/`true`/`false`

Setting this feature to `both` means that the profiler will automatically pick up context from either dd-trace-rb tracing or otel library; whereas setting it to `only` means that the profiler will skip dd-trace-rb tracing entirely and _only_ read context from the otel library.

(This was suggested by @alexjf in https://github.com/DataDog/dd-trace-rb/pull/3510#discussion_r1514692905, and also doubles as a way to raise confidence in releasing this since it allows us to have it off by default)

**Motivation:**

We have a customer interested in running this setup, so hopefully they'll be able to start using this feature.

**Additional Notes:**

**For this feature to work, otel traces need to be tagged with the dd-trace-rb runtime-id**. See below for an example of how to do this.

If the runtime-id is not set up, it will not be possible to see profiles scoped to individual requests in the datadog UX.

**How to test the change?**

On top of the added test coverage, I was able to see code hotspots working for the following sinatra example app:

```ruby
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'

  gem 'rackup'
  gem 'dogstatsd-ruby'
  gem 'datadog', git: 'https://github.com/datadog/dd-trace-rb', branch: 'ivoanjo/prof-9263-otlp-ruby-code-hotspots'
  gem 'sinatra'
  gem 'opentelemetry-api'
  gem 'opentelemetry-sdk'
  gem 'opentelemetry-instrumentation-sinatra'
  gem 'opentelemetry-exporter-otlp'
  gem 'pry'
end

require 'sinatra/base'
require 'opentelemetry/sdk'
require 'pry'

Datadog.configure do |c|
  c.service = 'ivoanjo-testing-opentelemetry-test'
  c.profiling.enabled = true
  c.profiling.advanced.preview_otel_context_enabled = :only
end

# Configure OpenTelemetry
OpenTelemetry::SDK.configure do |c|
  c.service_name = 'ivoanjo-testing-opentelemetry-test'
  c.use 'OpenTelemetry::Instrumentation::Sinatra'
end

class MyApp < Sinatra::Base
  get '/' do
    OpenTelemetry::Trace.current_span.add_attributes({'runtime-id' => Datadog::Core::Environment::Identity.id})
    sleep 1
    'Hello, OpenTelemetry!'
  end
end

MyApp.run!
```

And here's how it looks:

> ![image](https://github.com/user-attachments/assets/3f818e01-a8ce-4c61-a156-292ad11cab8f)

and 

> ![image](https://github.com/user-attachments/assets/701750b0-dcf4-4ea2-9976-b9e5fd72308f)
